DEVDOCS-2143 (Suggestions) (#682)
bigcommerce/dev-docs: docs/stencil-docs/page-builder/product-widget-tutorial.md

# Product Widget Tutorial
# Create Widgets Powered by GraphQL

<div class="otp" id="no-index">

### On this page
- [Prerequisites](#prerequisites)
- [Create a product widget](#create-a-product-widget)
- [Widget preview](#widget-preview)
- [Create the widget template](#create-the-widget-template)
- [Place the widget using Page Builder](#place-the-widget-using-page-builder)
- [Place the widget using the API](#place-the-widget-using-the-api)
- [Resources](#resources)

</div>

> Widgets are configurable and reusable components of content that merchants can display on their storefront. Widgets consist of a combination of HTML, JavaScript, CSS, and Handlebars, and are rendered as part of the storefront’s HTML. 

Reviewer: I would leave this sentence in the tutorial. I think it gives developers a quick overview of BC's definition of a widget without overcrowding the document.

Author: If an "understanding of widgets" is a prerequisite for the tutorial, why do we need to define it in the tutorial?

> * API OAuth token with the OAuth Scopes content set to `modify`.

Reviewer: ```suggestion
* API OAuth token with the OAuth Scopes content set to `modify`. For information on how to create an API account, see [Creating an API Account](https://support.bigcommerce.com/s/article/Store-API-Accounts#creating).
```

* Understanding of [widgets](https://developer.bigcommerce.com/api-docs/store-management/widgets/overview#widgets) and the [Widgets API](https://developer.bigcommerce.com/api-docs/store-management/widgets/overview).

> * Familiarity with [Page Builder](https://developer.bigcommerce.com/stencil-docs/page-builder/page-builder-overview)

Reviewer: ```suggestion
* Familiarity with [Page Builder](https://developer.bigcommerce.com/stencil-docs/page-builder/page-builder-overview).
```

## Create the widget template

> To [create the custom widget's template](https://developer.bigcommerce.com/api-reference/store-management/widgets/widget-template/createwidgettemplate), send a `POST` request to `/v3/content/widget-templates`.

Reviewer: ```suggestion
To [create the widget template](https://developer.bigcommerce.com/api-reference/store-management/widgets/widget-template/createwidgettemplate), send a `POST` request to `/v3/content/widget-templates`.
```

  "template": "<div style=\"text-align:center\">\n<h1>{{_.data.site.product.name}}</h1>\n<div>\n<img src=\"{{_.data.site.product.defaultImage.url}}\">\n</div>\n<div>\n<p>${{_.data.site.product.prices.price.value}}</p>\n</div>\n</div>"
}
```

[![Open in Request Runner](https://storage.googleapis.com/bigcommerce-production-dev-center/images/Open-Request-Runner.svg)](https://developer.bigcommerce.com/api-reference/store-management/widgets/widget-template/createwidgettemplate#requestrunner)

**[Response:](https://developer.bigcommerce.com/api-reference/store-management/widgets/widget-template/createwidgettemplate#responses)**

```json
{
  "data": {
    "uuid": "479e6ffe-73bc-4a0c-8e49-654d97691e8b",
    "name": "Product Widget",
    ...
}
```

|Property|Type|Description|
|-|-|-|

> |`name`|string|Name of the widget|

Reviewer: ```suggestion
|`name`|string|The name of the widget.|
```

> |`template`|string|The [widget template](https://developer.bigcommerce.com/api-docs/store-management/widgets/overview#widget-templates)|

Reviewer: ```suggestion
|`template`|string|The [widget template](https://developer.bigcommerce.com/api-docs/store-management/widgets/overview#widget-templates) rendered as Handlebars HTML.|
```

Author: Since we're linking to the widget template section in the widget overview, do we really need to include "rendered as handlebars HTML"?

Reviewer: I found that clarification helpful but I will leave this for Uwana to make the final determination.

Reviewer 2: @ektravel, I agree with @aglensmith that mentioning Handlebars immediately after the link is redundant since the first sentence on the linked page talks about Handlebars. We can remove that portion of the sentence.

> |`schema`|object|The widget settings [schema](https://developer.bigcommerce.com/stencil-docs/page-builder/widget-ui-schema) for [Page Builder](https://support.bigcommerce.com/s/article/Page-Builder) UI|

Reviewer: ```suggestion
|`schema`|object|The customizable widget settings [schema](https://developer.bigcommerce.com/stencil-docs/page-builder/widget-ui-schema) for [Page Builder](https://support.bigcommerce.com/s/article/Page-Builder) UI.|
```

Author: Are there widgets that aren't customizable?

Reviewer: I reckon that all BigCommerce widgets are customizable. We can remove all reference to "custom" and "customizable" from this tutorial.

> |`schema`|object|The widget settings [schema](https://developer.bigcommerce.com/stencil-docs/page-builder/widget-ui-schema) for [Page Builder](https://support.bigcommerce.com/s/article/Page-Builder) UI|

Reviewer 2: ```suggestion
|`schema`|object|The widget settings JSON [schema](https://developer.bigcommerce.com/stencil-docs/page-builder/widget-ui-schema) for [Page Builder](https://support.bigcommerce.com/s/article/Page-Builder) UI|
```
Again, for those reading through without clicking on every link.

> |`storefront_api_query`|string|[GraphQL Storefront API](https://developer.bigcommerce.com/api-docs/storefront/graphql/graphql-storefront-api-overview) query that provides widget data; accessed in template via `{{_.data}}` |

Reviewer: ```suggestion
|`storefront_api_query`|string|[GraphQL Storefront API](https://developer.bigcommerce.com/api-docs/storefront/graphql/graphql-storefront-api-overview) query that provides widget data; accessed in the template using `{{_.data}}`. |
```

## Place the widget using Page Builder

After [creating the widget template](#create-the-widget-template), you should see the widget listed in [Page Builder](https://developer.bigcommerce.com/stencil-docs/page-builder/page-builder-overview) under **Custom**.

![Product widget preview](https://raw.githubusercontent.com/bigcommerce/dev-docs/master/assets/images/product-widget.png)

>  Drag and drop the widget onto the desired page; doing so [creates a widget](https://developer.bigcommerce.com/api-reference/store-management/widgets/widget/createwidget) and a [widget placement](https://developer.bigcommerce.com/api-reference/store-management/widgets/placement/createplacement).

Reviewer: ```suggestion
 Drag and drop the widget onto the desired page; doing so creates [a widget](https://developer.bigcommerce.com/api-reference/store-management/widgets/widget/createwidget) and a [widget placement](https://developer.bigcommerce.com/api-reference/store-management/widgets/placement/createplacement).
```

> To see the placement and widget you just created, send a `GET` request to `/v3/content/placements`.

Reviewer: ```suggestion
To see the widget placement you just created, send a `GET` request to `/v3/content/placements`.
```

```http
GET https://api.bigcommerce.com/stores/{{STORE_HASH}}/v3/content/placements
X-Auth-Token: {{ACCESS_TOKEN}}
Accept: application/json
```

[![Open in Request Runner](https://storage.googleapis.com/bigcommerce-production-dev-center/images/Open-Request-Runner.svg)](https://developer.bigcommerce.com/api-reference/store-management/widgets/placement/getplacements#requestrunner)

**Response:**

```json
{
  "data": [
    {
      "uuid": "bb34b23b-0d4b-4b9b-9e24-c8b0dcfd5e08",
      "template_file": "pages/category",
      "region": "",
      "sort_order": 0,
      "entity_id": "21",
      "status": "active",
      "widget": {...},
      ...
  ],
  "meta": {...}
}
```

<div class="HubBlock--callout">
<div class="CalloutBlock--info">
<div class="HubBlock-content">

> ### Note
> * For more information on placing and configuring widgets in the control panel, see [Page Builder](https://support.bigcommerce.com/s/article/Page-Builder) in the Help Center.

</div>
</div>
</div>

> ## Place the widget using the API

Reviewer: Although helpful, I am not sure we need to include this section in this tutorial as it covers the steps available in the existing [Widget Tutorial](https://developer.bigcommerce.com/api-docs/store-management/widgets/tutorial). Perhaps, we can improve the Widget Tutorial and link to it instead?
```suggestion
## Place the widget using the API
```

Author: Since the tutorial is "Create a Featured Product Widget with GraphQL", I think we should include request runners for all API requests necessary to accomplish that. That's my thought process. That way, the reader can experiment with all the example requests using the request runners without leaving the page.

Reviewer: Understandable. I will leave this for Uwana to make the final determination.

Reviewer 2: @aglensmith @ektravel. You both have valid points. I'm concerned about the maintainability of having the same information in two places. This requires updating multiple places. Is there a way to rearrange the steps in the [Widget Tutorial](https://developer.bigcommerce.com/api-docs/store-management/widgets/tutorial) so that we can request that they follow those steps first, then complete a unique last step? This would be beneficial from a maintenance perspective, and since this is the very end of this tutorial. Is the sequence of steps too different to accomplish this?
